docs/architecture/adr-067-mempool-refactor.md

```diff
@@ -0,0 +1,137 @@
+# ADR 067: Mempool Refactor
```

Comment:
Being this is a mempool refactor, should we explore different ways of propagation (raptor codes and other erasure coding)?  @ValarDragon, do you think its worth discussing this?

Reply:
I think we should strive for doing that, but most erasure encoding techniques are really at the block-propogation level, merely with some optimistic early-encoding optimizations done within the mempool. So the way to proceed there would be to first architect block gossip to allow it (and changing the PartSet things), and then come back and work on adding it to the mempool.

(I also view this ADR as getting something done initially that handles sorting by fee, and sets up some external API's that will work for a bit. I still think we are going to want to switch to better back-end data structures ala #2147)

Reply:
I think that p2p layer improvements for selecting what txs to gossip, and ideal selections for gossiping can come later as well.